client/client.ml

```diff
@@ -14,37 +14,50 @@ let main ic oc =
       | Beginning -> Lwt.fail Exit
       | Established -> go (run context (`Hello Domain.(Domain [ "bar"; "com" ])))
       | Connected -> go (run context `Auth1)
-      | Auth1 -> go (run context `Auth2)
-      | Auth2 -> go (run context `Auth3)
-      | Auth3 -> go (run context (`Mail (Some { Path.local= `String "foo.bar.foo" ; domain= Domain.(Domain [ "laposte"; "net" ]) ; rest= [] }, [])))
-      | Mail -> go (run context (`Recipient Forward_path.(Forward_path { Path.local= `String "charles-edouard.lecat" ; domain= Domain.(Domain [ "epitech"; "eu" ]) ; rest= [] }, [])))
+      | Auth1 -> go (run context (`Auth2 login))
+      | Auth2 -> go (run context (`Auth3 password))
+      | Auth3 -> go (run context (`Mail sender))
+      | Mail -> go (run context (`Recipient recipient))
       | Rcpt -> go (run context `Data)
-      | Data -> go (run context (`Data_feed
-                  [ "Return-Path: foo.bar.foo@laposte.net"
-                  ; "Date: Fri, 9 Nov 2018 16:27:30 +0100 (CET)"
-                  ; "From: foo.bar.foo@laposte.net"
-                  ; "To: charles-edouard.lecat@epitech.eu"
-                  ; "Subject: Far from the Core, close to the Sky"
-                  ; "MIME-Version: 1.0"
-                  ; "Content-type: text/plain; charset=us-ascii"
-                  ; "Hello OCaml world !"
-                  ; "."
-                  ; ""]))
+      | Data -> go (run context (`Data_feed data))
       | Data_feed -> go (run context `Quit)
       | End -> Lwt.return_unit
   in
     go (run context `Establishment)
 
-let tls_wrap () =
-  let port = 465 in
-  let host = "smtp.laposte.net" in
+let tls_wrap server =
   Fmt.pf Fmt.stdout "Initializing TLS connection\n%!";
-  X509_lwt.authenticator `No_authentication_I'M_STUPID >>= fun authenticator ->
-  Tls_lwt.connect_ext
-    Tls.Config.(client ~authenticator ~ciphers:Ciphers.supported ())
-    (host, port) >>= fun (ic, oc) ->
-  Fmt.pf Fmt.stdout "Initialized TLS connection\n%!";
-  main ic oc
+  X509_lwt.authenticator `No_authentication_I'M_STUPID
+  >>= fun authenticator ->
+  let config = Tls.Config.(client ~authenticator ~ciphers:Ciphers.supported ()) in
+  Tls_lwt.connect_ext config server
 
-let () =
-  Lwt_main.run (tls_wrap ())
+(** Send an email
+    [~auth] is the tuple (login, password) base64-encoded, padded *)
+let send_mail
+    ~server
+    ~auth
+    ~from:(from_name, from_email)
+    ~to_:(to_name, to_email)
+    ?(headers=[])
+    ?(content="text/html")
```

Comment:
Is this argument a good idea ? It will definitively clash when using Mr.MIME. I think `headers` are fine, since they are optional and empty by default.
Also the `body` is a simple string, is this OK for Mr.MIME?

Reply:
Mr. MIME should be fully _streamable_ - main encoder about header was done in this way and about body, you can have a low-level view on [pecu](https://github.com/mirage/pecu.git) (quoted-printable) and [base64.rfc2045](https://github.com/mirage/ocaml-base64.git) (base64). So it should be fine - where Mr. MIME stills is a low-level API. I need to provide a way to make an header in OCaml easily and then, we will be able to make the glue between your code and Mr. MIME.

Reply:
Then, about the content-type, you can see [this API](https://github.com/mirage/mrmime/blob/master/lib/content_type.mli) where an `of_string : string -> (t, [ `Msg of string ]) result` and a `v : string -> t` for `Type.t` and `Subtype.t` should be ok (they needs to respect some assertions however).

Reply:
I removed the `content` argument and change the `body` argument to be a `unit -> 'a option`-style stream.
I'm not fully convinced this is the best API but it is definitively better than a `string`.
I hope this is enough to be able to plug Mr.MIME on top of this function later.

For base64, I think it would be better if this function did the encoding of the `login` and `password` arguments. I'll implement this later.